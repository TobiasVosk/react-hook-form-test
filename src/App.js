import './App.css';
import { useForm } from 'react-hook-form';
import React from 'react';
import { Button, Card } from 'react-bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css'

function App() {
  const { register, handleSubmit, watch, errors } = useForm();
  const [ submitted, setSubmitted ] = React.useState(false);
  const onSubmit = data => {
      setSubmitted(true)
      console.log(data);
  };

  return (
    <div className="container mt-4">
      <div className="card w-50 mx-auto">
        <div className="card-body">
          <form className="form-group" onSubmit={handleSubmit(onSubmit)}>
            <div>
              <input placeholder="Name" className="form-control mb-3" name="name" defaultValue="test" ref={register} title="name"/>

              <div className="mb-3">
                <input placeholder="Last Name" className={`form-control ${errors.lastName ? "invalid" : ""}`} name="lastName" title="lastName"
                  ref={register({
                    required: {value: true, message: "Last name is required"}, 
                    pattern: {value: /^[a-zA-Z ]+$/i, message: "Name is invalid. Only letters and spaces are accepted"}
                  })} />
                {errors.lastName && <small className="text-danger" title="lastNameValidation">{errors.lastName.message}</small>}
              </div>
              <Button className="btn btn-primary btn-block mb-3" type="submit" title="submitButton">Submit</Button>

              {submitted && <small className="text-success" title="submittedConfirmation">Submitted</small>}
            </div>
          </form>
        </div>
      </div>
    </div>
  );
}

export default App;
