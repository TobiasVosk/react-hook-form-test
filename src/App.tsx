import './App.css';
import 'bootstrap/dist/css/bootstrap.min.css'
import React from 'react';
import {BrowserRouter as Router, Route} from "react-router-dom"
import Form from './form';

function App() {

  return (
    <Router>
      <Route path="" component={Form}></Route>
    </Router>
  );
}

export default App;
