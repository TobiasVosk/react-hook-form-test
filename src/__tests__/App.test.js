import { act, fireEvent, getByTitle, render } from '@testing-library/react';
import App from '../App';

jest.mock('../submitUser');

describe("Form", () => {
  it("Should match the snapshot", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  })
})

describe("Name input", () => {
  it("can be written on", () => {
    const { queryByTitle } = render(<App />);
    const nameInput = queryByTitle("name");
    expect(nameInput.value).not.toBe("Changed the value");
    fireEvent.change(nameInput, { target: { value: "Changed the value" } })
    expect(nameInput.value).toBe("Changed the value");
  })
})

describe("Last name input", () => {
  it("can be written on", () => {
    const { queryByTitle } = render(<App />);
    const lastNameInput = queryByTitle("lastName");
    expect(lastNameInput.value).not.toBe("Changed the value");
    fireEvent.change(lastNameInput, { target: { value: "Changed the value" } })
    expect(lastNameInput.value).toBe("Changed the value");
  })
})

describe("Select input", () => {
  it("can be changed", () => {
    const { queryByTestId } = render(<App />);
    const categorySelect = queryByTestId("select-category");
    expect(categorySelect.value).toBe("");
    fireEvent.change(categorySelect, { target: { value: "A" } })
    expect(categorySelect.value).toBe("A");
  })
})

describe("Submit button", () => {
  it("should be disabled if no input is dirty", async () => {
    const { queryByTestId } = render(<App />);
    const submit = queryByTestId("submit-btn");

    expect(submit).toBeDisabled();
    await act(async () => {
      fireEvent.click(submit);
    });
  })

  it("should not submit if lastName is empty", async () => {
    const { queryByTitle, queryByTestId } = render(<App />);
    const submit = queryByTestId("submit-btn");

    const lastNameInput = queryByTitle("lastName");

    fireEvent.input(lastNameInput, { target: { value: "asd" } })
    fireEvent.input(lastNameInput, { target: { value: "" } })

    await act(async () => {
      fireEvent.click(submit);
    });

    const lastNameValidation = queryByTestId("last-name-validation-error")
    expect(lastNameValidation).toBeInTheDocument();
    expect(lastNameValidation.textContent).toEqual('Last name is required');
    expect(queryByTitle('submittedConfirmation')).not.toBeInTheDocument();
  })

  it("should not submit if lastName has symbols", async () => {
    const { queryByTitle, queryByTestId } = render(<App />);
    const submit = queryByTestId("submit-btn");

    const lastNameInput = queryByTitle("lastName");

    fireEvent.input(lastNameInput, { target: { value: "@" } })

    await act(async () => {
      fireEvent.click(submit);
    });

    const lastNameValidation = queryByTestId("last-name-validation-error")
    expect(lastNameValidation).toBeInTheDocument();
    expect(lastNameValidation.textContent).toEqual('Last name is invalid. Only letters and spaces are accepted');
    expect(queryByTitle('submittedConfirmation')).not.toBeInTheDocument();
  })

  it("should submit if lastName has letters and spaces, name is changed and select is empty", async () => {
    const { queryByTitle, queryByTestId } = render(<App />);
    const submit = queryByTestId("submit-btn");

    const name = "This is a wonderful name"
    const lastName = "This is a wonderful last name"

    const nameInput = queryByTitle("name");
    fireEvent.input(nameInput, { target: { value: name } })

    const lastNameInput = queryByTitle("lastName");
    fireEvent.input(lastNameInput, { target: { value: lastName } })

    expect(submit).not.toBeDisabled();

    await act(async () => {
      fireEvent.click(submit);
    });

    expect(queryByTitle('lastNameValidation')).not.toBeInTheDocument();
    expect(queryByTitle('submittedConfirmation')).toBeInTheDocument();

    const newNameInput = queryByTitle("name");
    expect(newNameInput.value).toEqual(name);

    const newLastNameInput = queryByTitle("lastName");
    expect(newLastNameInput.value).toEqual(lastName);

    const submittedLastName = queryByTestId("submitted-last-name");
    expect(submittedLastName).toBeInTheDocument();
    expect(submittedLastName.textContent).toEqual(lastName);

    const submittedName = queryByTestId("submitted-name")
    expect(submittedName).toBeInTheDocument();
    expect(submittedName.textContent).toEqual(name);

  })

  it("should submit if lastName has letters and spaces, name is changed and select is changed", async () => {
    const { queryByTitle, queryByTestId } = render(<App />);
    const submit = queryByTestId("submit-btn");

    const name = "This is a wonderful name"
    const lastName = "This is a wonderful last name"

    const nameInput = queryByTitle("name");
    fireEvent.input(nameInput, { target: { value: name } })

    const lastNameInput = queryByTitle("lastName");
    fireEvent.input(lastNameInput, { target: { value: lastName } })

    const categorySelect = queryByTestId("select-category");
    fireEvent.change(categorySelect, { target: { value: "A" } })

    expect(submit).not.toBeDisabled();

    await act(async () => {
      fireEvent.click(submit);
    });

    expect(queryByTitle('lastNameValidation')).not.toBeInTheDocument();
    expect(queryByTitle('submittedConfirmation')).toBeInTheDocument();

    const newNameInput = queryByTitle("name");
    expect(newNameInput.value).toEqual(name);

    const newLastNameInput = queryByTitle("lastName");
    expect(newLastNameInput.value).toEqual(lastName);

    const submittedLastName = queryByTestId("submitted-last-name");
    expect(submittedLastName).toBeInTheDocument();
    expect(submittedLastName.textContent).toEqual(lastName);

    const submittedName = queryByTestId("submitted-name")
    expect(submittedName).toBeInTheDocument();
    expect(submittedName.textContent).toEqual(name);

  })

})