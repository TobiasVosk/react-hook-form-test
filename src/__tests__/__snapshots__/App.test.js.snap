// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form Should match the snapshot 1`] = `
<div>
  <div
    class="container mt-4"
  >
    <div
      class="card w-50 mx-auto"
    >
      <div
        class="card-body"
      >
        <form
          class="form-group"
        >
          <div>
            <input
              class="form-control mb-3"
              name="name"
              placeholder="Name"
              title="name"
              value="test"
            />
            <div
              class="mb-3"
            >
              <input
                class="form-control"
                name="lastName"
                placeholder="Last Name"
                title="lastName"
              />
            </div>
            <select
              class="form-control mb-3"
              data-testid="select-category"
              name="category"
            >
              <option
                value=""
              >
                Select...
              </option>
              <option
                value="A"
              >
                Category A
              </option>
              <option
                value="B"
              >
                Category B
              </option>
            </select>
            <button
              class="btn btn-primary btn-block mb-3 btn btn-primary"
              data-testid="submit-btn"
              disabled=""
              title="submitButton"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
